### DISCUSSION TITLE
Add `data-wp-key` attribute

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#53062] [Interactivity API]
Sometimes we are going to need a `key` prop in some tags to improve the client-side navigation diffing, but `key` is not a valid HTML attribute. I suggest adding `data-wp-key` and then copying its value to `props.key` during the creation of the virtual nodes.

**Comment by SantosGuillamot** [comment]
For reference, we needed a `key` prop to improve the diffing of the images in the [Movies demo](https://github.com/WordPress/wp-movies-demo/). We ended up [using the HTML Tag Processor to add a key attribute to all the images](https://github.com/WordPress/wp-movies-demo/blob/main/wpmovies.php#L129-L135), but it would be great to have that `data-wp-key` attribute.

**Reply by luisherranz** [reply]
You ended up implementing this in the movies demo, didn't you Mario? If so, could you add a link to the code for reference?

**Reply by SantosGuillamot** [reply]
Yes, I got it working as part of the View Transitions API experiment, combining these two pull requests: one in the Block Hydration Experiments ([link](https://github.com/WordPress/block-interactivity-experiments/pull/241)), and another in the Movies Demo ([link](https://github.com/WordPress/wp-movies-demo/pull/52)).

If I remember correctly, to support the `data-wp-key` attribute, this is the relevant code: [link](https://github.com/WordPress/block-interactivity-experiments/blob/420fce412343851b014d66a6795c6ed5f800f16c/src/runtime/hooks.js#L133-L134).

**Reply by luisherranz** [reply]
Thanks!!

**Comment by luisherranz** [comment]
Also implemented in Woo in this PR: https://github.com/woocommerce/woocommerce-blocks/pull/10604

**Comment by luisherranz** [comment]
Implemented in https://github.com/WordPress/gutenberg/pull/53844.

- https://github.com/WordPress/gutenberg/pull/53844

**Comment by luisherranz** [comment]
I wonder if we should also use `id` as the key: If the element has an `id` and doesn't have a `data-wp-key`, use the `id` as the key.

What do you think?

**Reply by SantosGuillamot** [reply]
It could make sense. It would enable the possibility to define a `data-wp-key` through the editor, through the anchor input field.

On the other hand, I didn't think about it too much, but could this create unexpected behaviors? I'm thinking about unexpected duplicated `data-wp-key`, or plugins relying on `data-wp-key` assuming they have been added consciously.

**Reply by luisherranz** [reply]
I guess it's possible but highly unlikely because keys only need to be unique among their siblings.

Anyway, it's just a possibility. We can keep it in mind, but we don't have to do it yet 🙂

**Comment by luisherranz** [comment]
I'm going to close this as finished. We will reopen if necessary.
